Registration form: an optional "Confirm password" field

lenacom opened the pull request:
Continuation of the pull request https://github.com/edx/edx-platform/pull/12226

**Background:** Users often misspell their passwords and can't login a 2d time

**Studio Updates:** None.

**LMS Updates:** an optional "Confirm password" can be shown in the registration form. If the field is present a copy of password must be entered in it.

**Manual test instructions:**
Change lms.env.json:
`"REGISTRATION_EXTRA_FIELDS": {
        "password_copy": "required", 
`
Open http://127.0.0.1:8000/register and try to enter different "Password" and "Confirm password" and see a validation message.

edx-webhook commented:
Thanks for the pull request, @lenacom! I've created [OSPR-1420](https://openedx.atlassian.net/browse/OSPR-1420) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

gsong commented:
@lenacom Can you add a unit test for this new feature? It would help us along with the PR review process.

lenacom commented:
I'm adding tests now, especially tests for third-party login are needed.

gsong commented:
jenkins run js

gsong commented:
@lenacom Thank you so much, appreciate your effort.

srpearce commented on common/djangoapps/student/forms.py:
```diff
@@ -187,6 +187,13 @@ def __init__(
+                    if field_value == "required" and data.get("password"):
+                        self.fields[field_name] = forms.CharField(
+                            error_messages={
+                                "required": _("Please confirm password.")
```
Please confirm password. --> Confirm password  (no period)

srpearce commented on common/djangoapps/student/tests/test_create_account.py:
```diff
@@ -460,6 +460,28 @@ def assert_password_error(expected_error):
+            self.assert_error(params, "password_copy", expected_error)
+
+        # Missing
+        assert_password_copy_error("Please confirm password.")
```
Please confirm password. --> Confirm password (no period)

srpearce commented on openedx/core/djangoapps/user_api/views.py:
```diff
@@ -478,6 +486,36 @@ def _add_password_field(self, form_desc, required=True):
+        # meant to hold the user's retyped password.
+        password_copy_label = _(u"Confirm password")
+
+        error_msg = _(u"Please confirm password.")
```
Please confirm password. --> Confirm your password.

shaunagm commented:
Hi @lenacom.  To clarify, since you have opened a new PR: have you addressed the feedback from reviewers [in the previous thread](https://github.com/edx/edx-platform/pull/12226), or are you still working on that?

lenacom commented:
@shaunagm, yes, I addressed the feedback from the previous pull request. There was a comment that deprecated registration views shouldn't be updated but the old views are still used with microsites and our system uses microsites so we need this change in edx old registrations views.

shaunagm commented:
@lenacom can you address @srpearce's suggestions and also get that last test passing?  You might just need a rebase.

@marcotuts once the above are fixed, can this be moved to awaiting prioritization, or is there more product review to be done?

lenacom commented:
jenkins run bokchoy

nedbat commented:
@marcotuts Where do we stand on this?

gsong commented:
Pinged product again to see about the decision to review this PR.

gsong commented:
@lenacom Product has tentatively OKed this change.

* [ ] @lenacom Can you rebase to resolve conflicts?
* [ ] @cptvitamin Can you give a thumbs up from an accessibility perspective?

cptvitamin commented on lms/templates/register-form.html:
```diff
@@ -104,6 +104,12 @@ <h2 class="sr">${_('Required Information')}</h2>
+    % if settings.REGISTRATION_EXTRA_FIELDS.get('password_copy') == 'required':
+    <li class="field ${settings.REGISTRATION_EXTRA_FIELDS['password_copy']} text" id="field-password_copy">
+      <label for="city">${_('Confirm password')}</label>
+      <input id="password_copy" type="password" name="password_copy" value="" required aria-required="true" />
```
FYI, the ```aria-required```  attribute here (and on line 105 above) is unnecessary. The```required```  attribute is sufficient for all users. ```aria-required```  is only necessary when building a custom component and enforcing the required state of it via scripting. ```aria-required```  would be necessary in this case to inform the accessibiltiy API of this requirement so it could report it to the SR user. It's not doing any harm in this case, but its generally recommended not to have redundant attributes. This becomes more of a problem when attribute values are toggled, which is not the case here.

gsong commented:
Product has [given a go-ahead](https://openedx.atlassian.net/browse/OSPR-1420?focusedCommentId=242550&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-242550) on this.

lenacom commented:
I resolved a conflict with master and fixed a comment (removed "aria-required").

gsong commented:
@rock345 What do we need to do to move this into review?

schenedx commented on common/djangoapps/student/forms.py:
```diff
@@ -237,6 +244,16 @@ def clean_password(self):
         return password
 
+    def clean_password_copy(self):
+        """Enforce password_copy policies (if applicable)"""
```
`Enforce password confirmation policy (if applicable)`

schenedx commented on openedx/core/djangoapps/user_api/views.py:
```diff
@@ -341,6 +342,13 @@ def post(self, request):
             data["terms_of_service"] = data["honor_code"]
+        password = data.get("password")
+        password_copy = data.get("password_copy")
+        if password and password_copy and password != password_copy:
```
How come I don't see this check is gated by the setting `REGISTRATION_EXTRA_FIELDS['password_copy']=='required'`? Does this mean, the password confirmation logic always applies in the new registration view?

cptvitamin commented on lms/templates/register-form.html:
```diff
@@ -104,6 +104,12 @@ <h2 class="sr">${_('Required Information')}</h2>
     </li>
+    % if settings.REGISTRATION_EXTRA_FIELDS.get('password_copy') == 'required':
+    <li class="field ${settings.REGISTRATION_EXTRA_FIELDS['password_copy']} text" id="field-password_copy">
+      <label for="city">${_('Confirm password')}</label>
```
This says ```for="city"```  but I believe it should be ```for="password_copy"```

cptvitamin commented on lms/templates/register-form.html:
```diff
@@ -104,6 +104,12 @@ <h2 class="sr">${_('Required Information')}</h2>
       <label for="password">${_('Password')}</label>
       <input id="password" type="password" name="password" value="" required aria-required="true" />
```
I know this is beyond the scope of this PR, but while we are in there, can we get rid of the ```aria-required="true"``` attr here on this line?  It is redundant and unnecessary.  The ```required``` attribute works for all user agents that support it, including screen readers.  explicitly setting this value for screen readers is unnecessary.

edx-webhook commented:
Hello @lenacom: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.
